src/tye/CleanupHost.cs

```diff
@@ -0,0 +1,220 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.CommandLine;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using k8s;
+using k8s.Models;
+using Microsoft.Rest;
+using Microsoft.Tye.ConfigModel;
+using Newtonsoft.Json;
+
+namespace Microsoft.Tye
+{
+    public static class CleanupHost
+    {
+        public static async Task CleanupAsync(IConsole console, FileInfo path, Verbosity verbosity, bool interactive, bool whatIf)
+        {
+            var output = new OutputContext(console, verbosity);
+
+            output.WriteInfoLine("Loading Application Details...");
+
+            // We don't need to know anything about the services, just the application name.
+            var application = ConfigFactory.FromFile(path);
+
+            await ExecuteCleanupAsync(output, application, interactive, whatIf);
+        }
+
+        public static async Task ExecuteCleanupAsync(OutputContext output, ConfigApplication application, bool interactive, bool whatIf)
+        {
+            var config = KubernetesClientConfiguration.BuildDefaultConfig();
+
+            // Workaround for https://github.com/kubernetes-client/csharp/issues/372
+            var store = await KubernetesClientConfiguration.LoadKubeConfigAsync();
+            var context = store.Contexts.Where(c => c.Name == config.CurrentContext).FirstOrDefault();
+            config.Namespace ??= context?.ContextDetails?.Namespace;
+
+            var kubernetes = new Kubernetes(config);
+
+            // Due to some limitations in the k8s SDK we currently have a hardcoded list of resource
+            // types that we handle deletes for. If we start adding extensibility for the *kinds* of
+            // k8s resources we create, or the ability to deploy additional files along with the
+            // resources we understand then we should revisit this.
+            //
+            // Basically the challenges are:
+            //
+            // - kubectl api-resources --all (and similar) are implemented client-side (n+1 problem)
```

Comment:
What is the n+1 problem? I'm unfamiliar with it.

Reply:
The [n+1 problem](https://stackoverflow.com/questions/97197/what-is-the-n1-selects-problem-in-orm-object-relational-mapping) is a problem that comes up a lot in data design - usually with databases.

Basically it's named for having to do a query to discover `N` and then doing `N` queries to resolve all of the results. Querying all resources in k8s requires that, you have to enumerate all of the *kinds of things*, and then enumerate all of the things *by kind*.